File: Dynamic_Programming/14_Sep_2019_fibonacci_memoization/14_Sep_2019_fibonacci.py
#Python programme for memoized version of nth fibonacci number

#functtion to calculate nth fibonacci number

def fib(n, lookup):
    #base case
    if n == 0 or n == 1:
        lookup[n] = n
    
    #if the value if not calculated previously then calculate it
    if lookup[n] is None:
        lookup[n] == fib(n-1, lookup) + fib(n-2, lookup)
    
    #return the value corresponding to the value of n
    return lookup[n]


#Driver program to test the above function:
def main():
    n = 34
    #Declaration of lookup table 
    # Handles till n = 100
    lookup = [None]*(101)
    print("fibonacci number is {}".format(fib(n, lookup)))

if __name__ == "__main__":
    main()

File: Dynamic_Programming/14_Sep_2019_fibonacci/14_Sep_2019_fibonacci.c
/*
a) Memoization (Top Down): The memoized program for a problem is similar to the recursive version with a small modification that it looks into a lookup table before computing solutions. We initialize a lookup array with all initial values as NIL. Whenever we need the solution to a subproblem, we first look into the lookup table. If the precomputed value is there then we return that value, otherwise, we calculate the value and put the result in the lookup table so that it can be reused later.
*/
//Following is the memoized version for nth Fibonacci Number.


/*C program for memoized version for nth fibonacci number*/

#include<stdio.h>
#define NIL -1
#define MAX 100

int lookup[MAX];

/*Function to initialize NIL values in lookup table*/

void _initialize()
{
    int i;
    for(i = 0; i < MAX; i++)
    {
        lookup[i] = NIL;
    }
}

/*function for nth fibonacci number*/
int fib(int n)
{
    if(lookup[n] == NIL)
    {
        if(n <= 1)
        {
            lookup[n] == n;
        }
        else
        {
            lookup[n]  = fib(n-1) + fib(n-2);
        }
        
    }

    return lookup[n];
}

int main()
{
    int n = 40;
    _initialize();
    printf("Fibonacci number is %d", fib(n));
    return 0;
}